firedrake/tsfc_interface.py

```diff
@@ -54,19 +54,23 @@
 
 
 def tsfc_compile_form_hashkey(form, prefix, parameters, interface, diagonal):
-    # Drop prefix as it's only used for naming
-    return default_parallel_hashkey(form.signature(), prefix, parameters, interface, diagonal)
+    return default_parallel_hashkey(
+        form.signature(),
+        prefix,
+        utils.tuplify(parameters),
+        type(interface).__name__,
```

Comment:
In PCPatch this was not hashable and so we would get different disk hashes on different ranks.

Reply:
Hashing `interface` actually amounts to hashing the `dont_split` coefficient tuple, and it might not be straightforward. I am planning to remove `interface` (in favour of directly passing `dont_split`) in my PR, so I think you can for now probably choose to not cache if `interface` was given.

Reply:
At least passing `dont_split` suffices for current usage.

Maybe:
```
if interface is None:
    tree = tsfc_compile_form(...)
else:
    tree = original_tsfc_compile_form(...)
```

Reply:
I don't really want to hack something together. What is the correct long-term fix in your opinion? If you have a PR that addresses the nastiness of passing `interface` around then could we get that in first? I would like this merged within the next couple of weeks but it isn't holding anything up.

Reply:
@ksagiyam @connorjward is this something that needs to be changed in this PR? Or is this something that can be updated when `interface` is actually removed?

If it doesn't need to be in this PR, then @connorjward is this ready to go? It would be really good to get this merged, it prevents (/gets around) crashes we're regularly seeing with asQ.

Reply:
Opening an issue is a good idea, harder to lose track of than a comment in code. Although you could add a comment in the code pointing to the relevant issue.

Reply:
No. This is not about whether your current solution with `type(interface).__name__` is best or not; it introduces a new bug around what you just fixed. Suppose you make `interface` with:
```
    interface = make_builder(dont_split=(f, ))
``` 
and call `compile_form`. If you then you make another one with:
```
    interface = make_builder(dont_split=(g, ))
```
and call `compile_form`, you will end up grabbing the wrong value from the cache.

Reply:
What has happened here is I have been caught being lazy. The `interface` kwarg is a bad bit of code design and I wanted to get my code to pass the tests without having to think about it too hard.

I have just pushed something that will work.

Reply:
Resolving as this is now a duplicate.